### PR TITLE
Jpamodelgen fixes for panache

### PR DESCRIPTION
**Pull request by FroMage**
Please @gavinking can you review this?

- Support ORM/HR+Panache types such as anything that extends `PanacheEntityBase` or `PanacheRepositoryBase` to make sure to use their session getter type, but never generate a DAO because Quarkus will call the generated static methods. Also support annotations on `native` methods.
- Support `Uni<Mutiny.Session>` session types
- Support session getters with a default implementation, in which case never store the session in a field but always call the getter. This is mostly because reactive sessions can't be stored in a CDI context and need to be obtained for each call.
- Fix update/delete method return types that have to return `Uni<Integer|Void>` for HR
- When ORM/HR+Panache is detected, for types that are _not_ Panache entities or repositories, default to the Quarkus way of obtaining a session unless there's a session getter declared. The default way being `@Inject EntityManager` for ORM and `SessionOperations.getSession()` for HR. Note that this doesn't respect non-default PU for ORM, but we can improve later, if we have an entity defined (for Panache, we can override the PU per entity), and meanwhile users can always define a session getter to override. HR/Panache doesn't support non-default PU so this is not an issue.

Also, do you want docs for the changes? Tests?

### CODE REVIEW

**Comment by gavinking** [comment]
OK, on it.

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java [hunk 1]
```diff
@@ -283,7 +293,18 @@ else if ( containsAnnotation( method, Constants.HQL, Constants.SQL, Constants.FI
...
+			if(!isPanacheType(element)) {
+				dao = true;
+				sessionType = addDaoConstructor( getter );
+			} else {
+				sessionType = getter.getReturnType().toString();
+			}
+		} else if(context.usesQuarkusOrm() || context.usesQuarkusReactive()) {
+			dao = true;
+			sessionType = setupQuarkusDaoConstructor();
+		}
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:307]: This logic is getting a bit messy, so could we extract a method and add some comments explaining the cases we have to handle here?

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java [hunk 2]
```diff
@@ -1014,11 +1110,17 @@ private void validateUpdateHql(
...
 		if ( returnType == null
-				|| returnType.getKind() != TypeKind.VOID
-				&& returnType.getKind() != TypeKind.INT ) {
+				|| (!reactive 
+						&& returnType.getKind() != TypeKind.VOID
+						&& returnType.getKind() != TypeKind.INT)
+				|| (reactive 
+						// for reactive calls, don't use the returnType param, which has been ununi-ed, we want to check the full one
+						&& !method.getReturnType().toString().equals(Constants.UNI_VOID)
+						&& !method.getReturnType().toString().equals(Constants.UNI_INTEGER)) ) {
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:1121]: probably makes sense to extract a method for this long boolean condition.

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/CriteriaFinderMethod.java [hunk 3]
```diff
@@ -69,9 +69,13 @@ public String getAttributeDeclarationString() {
 						.append(");");
 			}
 		}
+		String localSessionName = isReactiveSession() ? "_session" : sessionName;
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/CriteriaFinderMethod.java:72]: Not sure how keen I am on `_session`. It's not the worst thing on Earth, but up until this point I managed to keep the generated code looking quite human-written. Is the `_` really necessary here?

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/CriteriaFinderMethod.java [hunk 4]
```diff
@@ -69,9 +69,13 @@ public String getAttributeDeclarationString() {
...
+		if(isReactiveSession()) {
+			declaration.append("\n\treturn ").append(sessionName).append(".flatMap(").append(localSessionName).append(" -> {");
+		}
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/CriteriaFinderMethod.java:75]: I think our convention is to use `chain()` in preference to `flatMap()`.

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java [hunk 5]
```diff
@@ -301,28 +322,63 @@ else if ( containsAnnotation( method, Constants.HQL, Constants.SQL, Constants.FI
...
+
+	private boolean isPanacheType(TypeElement type) {
+		return isOrmPanacheType(type) || isReactivePanacheType(type);
+	}
+
+	private boolean isOrmPanacheType(TypeElement type) {
+		ProcessingEnvironment processingEnvironment = this.context.getProcessingEnvironment();
+		TypeElement panacheRepositorySuperType = processingEnvironment.getElementUtils().getTypeElement("io.quarkus.hibernate.orm.panache.PanacheRepositoryBase");
+		TypeElement panacheEntitySuperType = processingEnvironment.getElementUtils().getTypeElement("io.quarkus.hibernate.orm.panache.PanacheEntityBase");
+		if(panacheRepositorySuperType == null || panacheEntitySuperType == null) {
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:361]: Perhaps move these strings to the `Constants` class. It's not my personal preference at all, but I've been trying to be consistent with what the guy who came before me did.

**Review by gavinking — APPROVED** [review]
Looks good. My main concern is how can test this within Hibernate test suite? Can we add a test dependency to Panache?

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java (hunk 2 above)
Comment (FroMage) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:1121]: And also I'm quite ashamed of the type test 🤦

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/CriteriaFinderMethod.java (hunk 3 above)
Comment (FroMage) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/CriteriaFinderMethod.java:72]: No, I just want it to be a different name than the session `Uni`, I can find better.

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/CriteriaFinderMethod.java (hunk 4 above)
Comment (FroMage) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/CriteriaFinderMethod.java:75]: OK

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java (hunk 5 above)
Comment (FroMage) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:361]: Fair.

**Comment by FroMage** [comment]
I think I've treated all the remarks. But for tests, I now realise that the detection allows for either Quarkus ORM _or_ HR, but not both (because that is not supported and both flavours are mutually exclusive), so I'm not sure (especially with Gradle) how to declare a dependency on those libs.
Also, it appears that the tests such as `DaoTest` only checks that the metamodel is generated, not that it has this or that method or structure, am I missing something?

**Comment by gavinking** [comment]
>  so I'm not sure (especially with Gradle) how to declare a dependency on those libs.

ugh.

> Also, it appears that the tests such as `DaoTest` only checks that the metamodel is generated, not that it has this or that method or structure, am I missing something?

For now that's all it does indeed. We could probably improve that by adding some code which _calls_ the generated code. But the fact that it is accepted by the compiler is already a fairly demanding test, though not perfect of course.

**Comment by FroMage** [comment]
> > Also, it appears that the tests such as `DaoTest` only checks that the metamodel is generated, not that it has this or that method or structure, am I missing something?
> 
> For now that's all it does indeed. We could probably improve that by adding some code which _calls_ the generated code. But the fact that it is accepted by the compiler is already a fairly demanding test, though not perfect of course.

What's supposed to be the result of this test:

```java
public interface Dao {
    @HQL("select upper('Hibernate')")
    TypedQuery<String> getName();
}
```

Because it appears to result in:

```java
@StaticMetamodel(Dao.class)
@Generated("org.hibernate.jpamodelgen.JPAMetaModelEntityProcessor")
public abstract class Dao_ {
}
```

And… I mean… it does compile, sure… 😂 but is it really supposed to be empty? 🤔

**Comment by FroMage** [comment]
> so I'm not sure (especially with Gradle) how to declare a dependency on those libs.

I've figured it out thanks to Steve.

**Comment by sebersole** [comment]
Nice!

**Comment by gavinking** [comment]
> And… I mean… it does compile, sure… 😂 but is it really supposed to be empty? 🤔

Um, no, that does not look right. 😬

**Comment by FroMage** [comment]
> I've figured it out thanks to Steve

Famous Last Words 🤦

Apparently checkerframework is something that's required, but it doesn't automatically applies to my new source sets and I'm getting a compilation error that it's looking for `checker-qual.jar`.

**Comment by gavinking** [comment]
@FroMage what's the status of this?

**Comment by FroMage** [comment]
Blocked on the Gradle build of the tests that require quarkus, due to checkerframework :(
I can push what I have, but it doesn't build.

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java [hunk 6]
```diff
@@ -482,11 +586,16 @@ else if ( kind == TypeKind.DECLARED ) {
...
-				: returnType;
+	private static TypeMirror ununi(TypeMirror returnType) {
+		if(returnType.getKind() != TypeKind.DECLARED) {
+			return returnType;
+		}
+		DeclaredType declaredType = (DeclaredType) returnType;
+		final TypeElement typeElement = (TypeElement) declaredType.asElement();
+		if(typeElement.getQualifiedName().contentEquals(Constants.UNI)) {
+			returnType = declaredType.getTypeArguments().get(0);
+		}
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:597]:
```suggestion
		if ( returnType.getKind() != TypeKind.DECLARED ) {
			return returnType;
		}
		DeclaredType declaredType = (DeclaredType) returnType;
		final TypeElement typeElement = (TypeElement) declaredType.asElement();
		if ( typeElement.getQualifiedName().contentEquals(Constants.UNI) ) {
			returnType = declaredType.getTypeArguments().get(0);
		}
```

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/CriteriaFinderMethod.java [hunk 7]
```diff
@@ -69,9 +69,13 @@ public String getAttributeDeclarationString() {
 						.append(");");
 			}
 		}
+		String localSessionName = isReactiveSession() ? "resolvedSession" : sessionName;
+		if(isReactiveSession()) {
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/CriteriaFinderMethod.java:73]:
```suggestion
		if ( isReactiveSession() ) {
```

**Review by gavinking — CHANGES_REQUESTED** [review]

**Comment by FroMage** [comment]
You just had to update that project, @gavinking, you couldn't just leave it be for the 6 months it took me to do this PR 😂

Well, I've rebased, I'll sort out the checkstyle issues and add tests for the non-panache fixes as well as the panache stuff.

**Comment by gavinking** [comment]
Check style doesn't enforce it.

**Comment by FroMage** [comment]
Yeah, so, where does this info come from?

**Comment by gavinking** [comment]
From looking at the rest of the codebase.

**Comment by FroMage** [comment]
> Yeah, I have _zero_ excuse :(

**Comment by FroMage** [comment]
OK, rebased on master, reformatted, includes tests for every feature/commit, tests now run as part of regular tests, and checkstyle passes.

So, IMO the only remaining thing might be to document some of those things?

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java [hunk 8]
```diff
@@ -430,31 +495,72 @@ private String addDaoConstructor(@Nullable ExecutableElement method) {
...
-						sessionVariableName,
-						dataStore(),
-						context.addInjectAnnotation(),
-						context.addNonnullAnnotation(),
-						method != null,
-						jakartaDataRepository
-				)
-		);
+		
+		if( method == null || !method.isDefault() ) {
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:499]: Why was this needed?

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java [hunk 9]
```diff
@@ -754,13 +865,14 @@ private void addFinderMethod(
 			ExecutableElement method,
 			@Nullable TypeMirror returnType,
 			@Nullable TypeElement containerType) {
+		returnType = returnType != null ? ununi( returnType ) : null;
```
[the remark this answers is not in the document]
Comment (FroMage) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:868]: Because otherwise `Uni<Entity>` would not be a valid return type?

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java [hunk 10]
```diff
@@ -393,31 +392,97 @@ private void addDefaultConstructor() {
 		}
 		return null;
 	}
+	
+	private void setupSession() {
+		jakartaDataRepository = hasAnnotation( element, JD_REPOSITORY );
+		ExecutableElement getter = findSessionGetter( element );
+		if( getter != null ) {
+			// Never make a DAO for Panache subtypes
+			if( !isPanacheType( element ) ) {
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:401]:
```suggestion
			if ( !isPanacheType( element ) ) {
```

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java [hunk 11]
```diff
@@ -393,31 +392,97 @@ private void addDefaultConstructor() {
...
+				repository = true;
+				sessionType = addDaoConstructor( getter );
+			}
+			else {
+				// For Panache subtypes, we look at the session type, but no DAO, we want static methods
+				sessionType = getter.getReturnType().toString();
+			}
+		}
+		else if( element.getKind() == ElementKind.INTERFACE
```
Comment (gavinking) [inline tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:410]:
```suggestion
		else if ( element.getKind() == ElementKind.INTERFACE
```

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java (hunk 8 above)
Comment (FroMage) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:499]: To support session getters in default methods, as shown in test:

```java
public interface Foo {
 public default Uni<Mutiny.Session> mysession() {
  // custom way to get a session
 }
 @HQL…
}
```

Comment (gavinking) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:499]: Hrm OK, that's interesting. Need to document that since it's a new feature.

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (gavinking) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:868]: But it was handled down in the `DECLARED_TYPE` case. `Uni` is a `DECLARED_TYPE`.

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java (hunk 8 above)
Comment (FroMage) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:499]: This is in the PR description ;)

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (gavinking) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:868]: And I don't think we ever have `Uni<X[]>` as a legal return type, do we?

[the remark this answers is not in the document]
Comment (FroMage) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:868]: The first check was not: we want to check the type argument of the `Uni`.

[the remark this answers is not in the document]
Comment (FroMage) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:868]: > And I don't think we ever have `Uni<X[]>` as a legal return type, do we?

Why not? Any valid type `X` is also a valid `Uni<X>` type, no?

File: tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java (hunk 8 above)
Comment (FroMage) [reply tooling/metamodel-generator/src/main/java/org/hibernate/jpamodelgen/annotation/AnnotationMetaEntity.java:499]: Lucky for us it's still there: https://github.com/hibernate/hibernate-orm/pull/7599#issue-2025938172 :)

Let me know if you want me to update the docs.

**Review by gavinking — APPROVED** [review]

**Comment by FroMage** [comment]
I can rebase again, but we should really merge this because I'm not going to keep doing that every day ;)
